internal/app/secretless/handlers/pg/protocol/ssl.go

```diff
@@ -0,0 +1,149 @@
+package protocol
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+)
+
+type options map[string]string
+
+type TLSConfigWrapper struct {
+	UseTLS bool
+	tls.Config
+	VerifyCaOnly bool
+	Options options
+}
+
+func ResolveTLSConfig(o options) (TLSConfigWrapper, error) {
+	tlsConf := TLSConfigWrapper{Options:o, UseTLS: true}
+
+	switch mode := o["sslmode"]; mode {
+	case "disable":
+		tlsConf.UseTLS = false
+		return tlsConf, nil
+		// "require" is the default.
+	case "", "require":
+		// We must skip TLS's own verification since it requires full
+		// verification since Go 1.3.
+		tlsConf.InsecureSkipVerify = true
+
+		// From http://www.postgresql.org/docs/current/static/libpq-ssl.html:
+		//
+		// Note: For backwards compatibility with earlier versions of
+		// PostgreSQL, if a root CA file exists, the behavior of
+		// sslmode=require will be the same as that of verify-ca, meaning the
+		// server certificate is validated against the CA. Relying on this
+		// behavior is discouraged, and applications that need certificate
+		// validation should always use verify-ca or verify-full.
+		if len(o["sslrootcert"]) > 0 {
+			tlsConf.VerifyCaOnly = true
+		}
+	case "verify-ca":
+		// We must skip TLS's own verification since it requires full
+		// verification since Go 1.3.
+		tlsConf.InsecureSkipVerify = true
+		tlsConf.VerifyCaOnly = true
+	//case "verify-full":
+	//	tlsConf.ServerName = o["host"]
+	default:
+		// TODO add verify-full below
+		return TLSConfigWrapper{}, fmt.Errorf(`unsupported sslmode %q; only "require" (default), "verify-ca", and "disable" supported`, mode)
+	}
+
+	return tlsConf, nil
+}
+
+// HandleSSLUpgrade upgrades a net.Conn using TLSConfigWrapper
+func HandleSSLUpgrade(connection net.Conn, tlsConf TLSConfigWrapper) (net.Conn, error) {
+	err := sslClientCertificates(&tlsConf.Config, tlsConf.Options)
+	if err != nil {
+		return nil, err
+	}
+	err = sslCertificateAuthority(&tlsConf.Config, tlsConf.Options)
+	if err != nil {
+		return nil, err
+	}
+
+	// Accept renegotiation requests initiated by the backend.
+	//
+	// Renegotiation was deprecated then removed from PostgreSQL 9.5, but
+	// the default configuration of older versions has it enabled. Redshift
+	// also initiates renegotiations and cannot be reconfigured.
+	// TODO: what does mysql require for this ?
+	tlsConf.Renegotiation = tls.RenegotiateFreelyAsClient
+
+	client := tls.Client(connection, &tlsConf.Config)
+	if tlsConf.VerifyCaOnly {
+		err := sslVerifyCertificateAuthority(client, &tlsConf.Config)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return client, nil
+}
+
+// sslClientCertificates adds the certificate specified in the "sslcert" and
+// "sslkey" settings
+func sslClientCertificates(tlsConf *tls.Config, o options) error {
```

Comment:
As is, `addClientCertsToConfiguration` would be a more accurate name, since this function mutates `tlsConf` rather than returning the certificates, as I'd expect given the current name and the convention of naming returning funcs after what the return.

That said, in this case there's no good reason to pass in `tlsConf` afaict (in contrast the the funcs below), so I'd consider keeping the current name, but having the func do nothing but _return_ the `[]tls.Certificate{cert}`.  The calling code could then do:
```go
tlsConf.Certificates = clientCertificates(options)
```